pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -1048,6 +1082,57 @@ func (m *ManagerImpl) GetDevices(podUID, containerName string) ResourceDeviceIns
 	return m.podDevices.getContainerDevices(podUID, containerName)
 }
 
+func (m *ManagerImpl) UpdateAllocatedResourcesStatus(pod *v1.Pod, status *v1.PodStatus) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	// Today we ignore edge cases that are not likely to happen:
+	//  - update statuses for containers that are in spec, but not in status
+	//  - update statuses for resources requested in spec, but with no information in podDevices
+	for i, containerStatus := range status.ContainerStatuses {
+		devices := m.podDevices.getContainerDevices(string(pod.UID), containerStatus.Name)
+
+		for resourceName, deviceInstances := range devices {
+			for id, d := range deviceInstances {
+				health := pluginapi.Healthy
+				// this is unlikely, but check for existence here anyways
+				if r, ok := m.allDevices[resourceName]; ok {
+					if _, ok := r[id]; ok {
+						health = m.allDevices[resourceName][id].Health
+					}
+				}
+
+				d.Health = health
+
+				deviceInstances[id] = d
```

Comment:
this seems is the simplest implementation and makes sense for alpha, but I'm concerned about the locking and the lock contention (xref my other comment about e2e test). I think an interesting alternative to try out is to add all the necessary data into `resourceupdate.Update` - we should have it when we generate the event, so `UpdateAllocatedResources` can be enabled to work only with local data, avoiding any other call to devicemanager. This has the potential to reduce the API surface, the lock contention and should be easier to test.

Reply:
so I think trying out this approach, or at least discussing it in depth, should be at least a beta blocker.

Reply:
OTOH because of the pod workers API (and doesn't make sense to change it for this flow IMO) is not easy or possible at all to pass this data, so this is not a easy fix. Let's keep it for beta.